File: haskell/day01.hs
import System.Environment ( getArgs )
import Data.List ( sort )

main :: IO ()
main = do
        args <- getArgs
        content <- readFile (head args)
        print (partOne content)
        print (partTwo content)

extractEven :: [String] -> [Int]
extractEven [] = []
extractEven (x:xs) = read x : extractEven (drop 1 xs)

extractOdd :: [String] -> [Int]
extractOdd xs = extractEven (drop 1 xs)

partOne :: String -> Int
partOne input =
        sum (zipWith (\x y -> abs (x - y))
                (sort (extractEven (words input)))
                (sort (extractOdd  (words input))))

count :: Int -> [Int] -> Int
count i xs = length (filter (== i) xs)

partTwo :: String -> Int
partTwo input = 
        sum (map (\x -> x * count x (extractOdd (words input)))
                (extractEven (words input)))

File: scala/day01.scala
//> using toolkit 0.6.0

object Day01:
	val path = os.pwd / ".." / "input" / "day01" / "input1.txt"
	val lines = os.read.lines(path).toVector
	val groups =
		lines
		.map(_.split(" ").filterNot(_.isEmpty).map(_.toInt))
		.map(xs => (xs(0), xs(1)))
		.unzip

	def partOne(): Unit =
		val (leftList, rightList) = (groups._1.sorted, groups._2.sorted)
		val pairs = leftList.zip(rightList)
		val distances = pairs.map((lhs, rhs) => (rhs - lhs).abs)
		val result = distances.sum
		println(result)

	def partTwo(): Unit = 
		val (leftList, rightList) = groups
		val similarityScore = leftList.map(e => e * rightList.count(_ == e))
		val result = similarityScore.sum
		println(result)

	@main def run(): Unit =
		partOne()
		partTwo()